### PR TITLE
Refactor fmt::Display and fmt::Debug impls in ppaux

### LINKED ISSUE #44887: Function pointer types never display higher-ranked region when they are used as argument types
```rust
fn f<F>(_: F) where F: for<'a> Fn(&'a u8, for<'b> fn(&'b u8)) {}

fn main() {
    f(|_: (), _: ()| {});
}
```

```
error[E0631]: type mismatch in closure arguments
 --> src/main.rs:4:5
  |
4 |     f(|_: (), _: ()| {});
  |     ^ ----------------- found signature of `fn((), ()) -> _`
  |     |
  |     expected signature of `for<'a> fn(&'a u8, fn(&'b u8)) -> _`
  |
  = note: required by `f`
```

Here, only `for<'b>` is omitted.

## Meta

```
$ rustc --version --verbose
rustc 1.22.0-nightly (930d3b17d 2017-09-26)
binary: rustc
commit-hash: 930d3b17dd91b6564cf535ac717c688db757be5d
commit-date: 2017-09-26
host: x86_64-unknown-linux-gnu
release: 1.22.0-nightly
LLVM version: 4.0
```

### PR DESCRIPTION
**Pull request by tirr-c**
Also fixes #44887.

There was a problem that unnamed late-bound regions are *always* named `'r` while they are displayed using `std::fmt::Display`.

---

```rust
fn main() {
    f(|_: (), _: ()| {});
}
fn f<F>(_: F) where F: Fn(&(), for<'r> fn(&'r ())) {}
```

Before (incorrectly shadows lifetime, `for<'r>` omitted for the second argument):

```
error[E0631]: type mismatch in closure arguments
 --> test.rs:2:5
  |
2 |     f(|_: (), _: ()| {});
  |     ^ ----------------- found signature of `fn((), ()) -> _`
  |     |
  |     expected signature of `for<'r> fn(&'r (), fn(&'r ())) -> _`
  |
  = note: required by `f`
```

After:

```
error[E0631]: type mismatch in closure arguments
 --> test.rs:2:5
  |
2 |     f(|_: (), _: ()| {});
  |     ^ ----------------- found signature of `fn((), ()) -> _`
  |     |
  |     expected signature of `for<'s> fn(&'s (), for<'r> fn(&'r ())) -> _`
  |
  = note: required by `f`
```

r? @nikomatsakis

### CODE REVIEW

File: src/librustc/ty/context.rs [hunk 1]
```diff
@@ -882,6 +882,10 @@ pub struct GlobalCtxt<'tcx> {
     /// Maps Expr NodeId's to `true` iff `&expr` can have 'static lifetime.
     pub rvalue_promotable_to_static: RefCell<NodeMap<bool>>,
 
+    pub display_used_late_bound_region_names: RefCell<Option<FxHashSet<Name>>>,
```
Comment (arielb1) [inline src/librustc/ty/context.rs]: This is likely to cause problems with incremental compilation. I think the amount of  globals ppaux is depending on is getting far out of hand, and we should stop using `Display` & `Debug` and start using our own trait.

**Comment by tirr-c** [comment]
I've refactored ppaux, and I think it's halfway done. I believe we can make this better, especially those macros; they are very rough. I defined macros because I can't just write `impl<T> fmt::Display for T where T: Print`.

r? @arielb1

**Comment by arielb1** [comment]
```
[00:35:42] ---- [ui] ui/regions-fn-subtyping-return-static.rs stdout ----
[00:35:42] 	normalized stderr:
[00:35:42] error[E0308]: mismatched types
[00:35:42]   --> $DIR/regions-fn-subtyping-return-static.rs:51:12
[00:35:42]    |
[00:35:42] 51 |     want_F(bar); //~ ERROR E0308
[00:35:42]    |            ^^^ expected concrete lifetime, found bound lifetime parameter 'cx
[00:35:42]    |
[00:35:42]    = note: expected type `for<'cx> fn(&'cx S) -> &'cx S`
[00:35:42]               found type `for<'a> fn(&'a S) -> &S {bar::<'_>}`
[00:35:42] 
[00:35:42] error: aborting due to previous error
[00:35:42] 
[00:35:42] 
[00:35:42] 
[00:35:42] expected stderr:
[00:35:42] error[E0308]: mismatched types
[00:35:42]   --> $DIR/regions-fn-subtyping-return-static.rs:51:12
[00:35:42]    |
[00:35:42] 51 |     want_F(bar); //~ ERROR E0308
[00:35:42]    |            ^^^ expected concrete lifetime, found bound lifetime parameter 'cx
[00:35:42]    |
[00:35:42]    = note: expected type `for<'cx> fn(&'cx S) -> &'cx S`
[00:35:42]               found type `fn(&'a S) -> &S {bar::<'_>}`
[00:35:42] 
[00:35:42] error: aborting due to previous error
[00:35:42] 
[00:35:42] 
[00:35:42] 
[00:35:42] diff of stderr:
[00:35:42] 
[00:35:42]  error[E0308]: mismatched types
[00:35:42]    --> $DIR/regions-fn-subtyping-return-static.rs:51:12
[00:35:42]     |
[00:35:42]  51 |     want_F(bar); //~ ERROR E0308
[00:35:42]     |            ^^^ expected concrete lifetime, found bound lifetime parameter 'cx
[00:35:42]     |
[00:35:42]     = note: expected type `for<'cx> fn(&'cx S) -> &'cx S`
[00:35:42] -              found type `fn(&'a S) -> &S {bar::<'_>}`
[00:35:42] +              found type `for<'a> fn(&'a S) -> &S {bar::<'_>}`
[00:35:42]  
[00:35:42]  error: aborting due to previous error
```

File: src/librustc/util/ppaux.rs [hunk 2]
```diff
@@ -32,310 +33,564 @@ use syntax::ast::CRATE_NODE_ID;
...
-    } else {
-        did.index = key.parent.unwrap_or_else(
-            || bug!("finding type for {:?}, encountered def-id {:?} with no parent",
-                    did, did));
-        parameterized(f, substs, did, projections)?;
-        return write!(f, "::{}", key.disambiguated_data.data.as_interned_str());
-    };
+default impl<T> DefaultWith<T> for PrintContext {}
+
+impl<'tcx, T> DefaultWith<T> for PrintContext
```
Comment (arielb1) [inline src/librustc/util/ppaux.rs]: that's an... ok use of specialization. I suppose we can avoid it if we have to?

Comment (tirr-c) [reply src/librustc/util/ppaux.rs]: Oops, I missed your review -- what do you mean by that? I can't quite get it...

**Comment by arielb1** [comment]
still same test needs updating

**Comment by arielb1** [comment]
Nice cleanup with the macros. There were a few directions I wanted to go there, but they can be in a separate PRs.

LGTM, r=me after this passes travis.

**Comment by tirr-c** [comment]
I haven't fully tested this, I'll let travis to do necessary tests.

File: src/librustc/util/ppaux.rs [hunk 3]
```diff
@@ -32,310 +33,543 @@ use syntax::ast::CRATE_NODE_ID;
...
+                is_verbose: is_verbose,
+                identify_regions: identify_regions,
+                used_region_names: None,
+                region_index: 0,
+                binder_depth: 0,
+            }
+        })
+    }
+    fn prepare_late_bound_region_info<'tcx, T>(&mut self, value: &ty::Binder<T>)
+    where T: TypeFoldable<'tcx>
```
Comment (arielb1) [inline src/librustc/util/ppaux.rs]: nit: indentation, should be

```Rust
    fn prepare_late_bound_region_info<'tcx, T>(&mut self, value: &ty::Binder<T>)
        where T: TypeFoldable<'tcx>
```

**Comment by arielb1** [comment]
nah you don't have to fix the indentation

@bors r+

**Comment by arielb1** [comment]
Thanks for the PR! That's a very cool use of macros.

**Comment by lunaperezclaudio-svg** [comment]
2464707387 ubicasion

**Pull request merged**